File: ocaml/fft_array.ml
(**************************************************************************************************
 Fast Fourier Transform -- OCaml Version
 This version implements Cooley-Tukey algorithm for powers of 2 only using OCaml arrays.

 José Alexandre Nalon
 **************************************************************************************************
 There are different ways to compile and run an OCaml program. To run directly from the command
 line, just type:

 $ ocaml fft_array.ml

 And the program will run. To compile the program, use the command:

 $ ocamlc -o fft_array fft_array.ml

 This will generate an executable file named `fft_array` in the same directory, that can be run
 from the command line by typing:

 $ ./fft_array
 **************************************************************************************************)

(**************************************************************************************************
 Open needed modules:
 **************************************************************************************************)
open Array                             (* Array of numbers; *)
open Complex                           (* Complex numbers; *)
open Printf                            (* Output results; *)
open Sys                               (* Time measurement; *)


(**************************************************************************************************
 Definitions:
 **************************************************************************************************)
let repeats = 50
let pi = 3.1415926535897931


(**************************************************************************************************
 Auxiliary Functions:
 **************************************************************************************************)
(* Convert integer to complex. *)
let complex_of_int n = { re=(float_of_int n); im=0. }


(* Integer power of a complex number. *)
let cpow w n = Complex.pow w (complex_of_int n)


(* Creates a sequence of integer numbers. *)
let range a b = List.init (b - a + 1) (fun n -> n + a)


(**************************************************************************************************
 Auxiliary function: complex_show
   Pretty printing of an array of complex numbers, used to inspect results.

 Parameters:
   x
     An array of complex numbers;
 **************************************************************************************************)
let complex_show x =
    for i = 0 to ((length x) - 1) do
        printf "( %7.4f, %7.4f )\n" x.(i).re x.(i).im
    done


(**************************************************************************************************
 Measure execution time through repeated calls to a (Fast) Fourier Transform function.

 Parameters:
   f
     Function to be called;
   size
     Power of two of the size of the vector on which the transform will be applied;
   repeats
     Number of times the function will be called.

 Returns:
   The average execution time for that function with a vector of the given size.
 **************************************************************************************************)
let time_it f size repeats =
    let x = Array.init size (fun n -> { re=(float n); im=0. }) in
    let t0 = Sys.time () in
    for i = 1 to size do
        f x;
    done;
    (Sys.time () -. t0) /. (float repeats)


(**************************************************************************************************
 Discrete Fourier Transform directly from the definition, an algorithm that has O(N^2) complexity.

 Parameters:
   x
     The vector of which the DFT will be computed. Given the nature of the implementation, there is
     no restriction on the size of the list, although it will almost always be called with a power
     of two size to give a fair comparison.

 Returns:
     A complex-number vector of the same size, with the coefficients of the DFT.
 **************************************************************************************************)
let direct_ft x =
    let nx = length x in
    let tX = Array.init nx (fun n -> Complex.zero) in
    let w = exp { re=0.; im= -2. *. pi /. (float nx) } in
    let wk = ref { re=1.; im=0. } in
    for k = 0 to (nx - 1) do
        let wkn = ref { re=1.; im=0. } in
        for n = 0 to (nx - 1) do
            tX.(k) <- add tX.(k) (mul !wkn x.(n));
            wkn := mul !wkn !wk
        done;
        wk := mul !wk w
    done;
    tX


(**************************************************************************************************
 Fast Fourier Transform using a recursive decimation in time algorithm. This has O(N log_2(N))
 complexity.

 Parameters:
   x
     The vector of which the DFT will be computed. This should have length power of 2;

 Returns:
   A complex-number vector of the same size, with the coefficients of the DFT.
 **************************************************************************************************)
let rec recursive_fft x =
    if Array.length x == 1 then
        x
    else
        let nx = length x in
        let n2 = nx / 2 in
        let xe = Array.make n2 Complex.zero in
        let xo = Array.make n2 Complex.zero in
        let tX = Array.make nx Complex.zero in
        for k = 0 to (n2 - 1) do
            xe.(k) <- x.(2*k);
            xo.(k) <- x.(2*k + 1)
        done;
        let tXe = recursive_fft xe in
        let tXo = recursive_fft xo in
        let w = exp { re=0.; im= -2. *. pi /. (float nx) } in
        let wk = ref { re=1.; im=0. } in
        for k = 0 to (n2 - 1) do
            let wkxo = mul !wk tXo.(k) in
            tX.(k) <- add tXe.(k) wkxo;
            tX.(k+n2) <- sub tXe.(k) wkxo;
            wk := mul !wk w
        done;
        tX


(**************************************************************************************************
 Bit-reversed version of an integer number.

 Parameters:
   k
     The number to be bit-reversed;
   r
     The number of bits to take into consideration when reversing.

 Returns:
   The number k, bit-reversed according to integers with r bits.
 **************************************************************************************************)
let bit_reverse k r =
    let l = ref 0 in
    let k = ref k in
    for i = 1 to r do
        l := 2 * !l + (!k mod 2);
        k := !k / 2
    done;
    !l


(**************************************************************************************************
 Fast Fourier Transform using an iterative in-place decimation in time algorithm. This has
 O(N log_2(N)) complexity, and since there are less function calls, it will probably be marginally
 faster than the recursive versions.

 Parameters:
   x
     The vector of which the FFT will be computed. This should always be called with a vector of a
     power of two length, or it will fail. No checks on this are made.

 Returns:
   A complex-number vector of the same size, with the coefficients of the DFT.
 **************************************************************************************************)
let iterative_fft x =
    let nx = length x in
    let tX = Array.make nx Complex.zero in
    let r = int_of_float ((Float.log (float nx)) /. (Float.log 2.)) in
    for k = 0 to (nx - 1) do
        let l = bit_reverse k r in
        tX.(l) <- x.(k);
    done;
    let step = ref 1 in
    for k = 0 to (r - 1) do
        let l = ref 0 in
        while !l < nx do
            let w = exp { re=0.; im= -. pi /. (float !step) } in
            let wkn = ref { re=1.; im=0. } in
            for n = 0 to (!step - 1) do
                let p = !l + n in
                let q = p + !step in
                tX.(q) <- sub tX.(p) (mul !wkn tX.(q));
                tX.(p) <- sub (mul { re=2.; im=0. } tX.(p)) tX.(q);
                wkn := mul !wkn w
            done;
            l := !l + 2 * !step
        done;
        step := !step * 2
    done;
    tX


(**************************************************************************************************
 Main function:
 **************************************************************************************************)
let main () =

    (* Start by printing the table with time comparisons: *)
    printf "+---------+---------+---------+---------+---------+---------+\n";
    printf "|    N    |   N^2   | N logN  | Direta  | Recurs. | Itera.  |\n";
    printf "+---------+---------+---------+---------+---------+---------+\n";

    (* Try it with vectors with size ranging from 32 to 1024 samples: *)
    for r = 5 to 10 do

        (* Compute the average execution time: *)
        let n = int_of_float (2. ** (float r)) in
        let dtime = time_it direct_ft n repeats in
        let rtime = time_it recursive_fft n repeats in
        let itime = time_it iterative_fft n repeats in

        (* Print the results: *)
        printf "| %7d | %7d | %7d | %7.4f | %7.4f | %7.4f |\n" n (n*n) (n*r) dtime rtime itime;
        flush stdout

    done;
    printf "+---------+---------+---------+---------+---------+---------+\n"


let () = main ()

File: python/fft_array.py
# -*- coding: utf-8 -*-
####################################################################################################
# Fast Fourier Transform -- Python 3 Version using native arrays.
# This version implements Cooley-Tukey algorithm for powers of 2 only.
#
# José Alexandre Nalon
####################################################################################################
# This is a function only library, and executing this file won't give any results. Notice however,
# that, since this implementation uses only standard Python 3 objects, you can try to include this
# in scripts and run them using Pypy 3, Cython 3 or any other Python 3 implementation.


####################################################################################################
# Import needed modules:
from array import array                        # Efficient arrays;
import math                                    # Math operations;


####################################################################################################
# Direct FT:
def direct_ft(x):
    """
    Discrete Fourier Transform directly from the definition, an algorithm that has O(N^2)
    complexity. This implementation uses arrays from the standard Python library.

    :Parameters:
      x
        The vector of which the DFT will be computed. Given the nature of the implementation, there
        is no restriction on the size of the vector, although it will almost always be called with
        a power of two size to give a fair comparison.

    :Returns:
      A complex-number vector of the same size, with the coefficients of the DFT.
    """
    N = len(x)                                 # Length of the vector;
    x_real = array('f', [ xr.real for xr in x ])
    x_imag = array('f', [ xi.imag for xi in x ])
    X_real = array('f', [ 0. ] * N)            # Accumulate the results;
    X_imag = array('f', [ 0. ] * N)
    W_real = math.cos(-2*math.pi/N)            # Twiddle factors;
    W_imag = math.sin(-2*math.pi/N)
    Wk_real = 1.
    Wk_imag = 0.
    for k in range(0, N):                      # Compute the kth coefficient;
        Wkn_real = 1.
        Wkn_imag = 0.
        for n in range(0, N):                  #   Operate the summation;
            X_real[k] += x_real[n]*Wkn_real - x_imag[n]*Wkn_imag       # Compute every term;
            X_imag[k] += x_real[n]*Wkn_imag + x_imag[n]*Wkn_real
            aux_real = Wkn_real * Wk_real - Wkn_imag * Wk_imag         # Update twiddle factors;
            aux_imag = Wkn_real * Wk_imag + Wkn_imag * Wk_real
            Wkn_real = aux_real
            Wkn_imag = aux_imag
        aux_real = Wk_real * W_real - Wk_imag * W_imag
        aux_imag = Wk_real * W_imag + Wk_imag * W_real
        Wk_real = aux_real
        Wk_imag = aux_imag
    return [ Xr + 1j*Xi for Xr, Xi in zip(X_real, X_imag) ]


####################################################################################################
# Recursive Decimation-in-time FFT:
def recursive_fft(x):
    """
    Fast Fourier Transform using a recursive decimation in time algorithm. This has O(N log_2(N))
    complexity. This implementation uses arrays from the standard Python library.

    :Parameters:
      x
        The vector of which the DFT will be computed. Given the nature of the implementation, there
        is no restriction on the size of the vector, although it will almost always be called with
        a power of two size to give a fair comparison.

    :Returns:
      A complex-number vector of the same size, with the coefficients of the DFT.
    """
    if len(x) == 1:                            # A length-1 vector is its own FT;
        return x
    else:
        N = len(x)                             # Length of the vector;

        # Separated transform of even and odd samples:
        Xe = recursive_fft(x[0::2])
        Xo = recursive_fft(x[1::2])
        Xe_real = array('f', [ xr.real for xr in Xe ])
        Xe_imag = array('f', [ xi.imag for xi in Xe ])
        Xo_real = array('f', [ xr.real for xr in Xo ])
        Xo_imag = array('f', [ xi.imag for xi in Xo ])

        # Twiddle factors:
        kn = list(range(0, N//2))
        W_real = array('f', [ math.cos(-2*math.pi*k/N) for k in kn ])
        W_imag = array('f', [ math.sin(-2*math.pi*k/N) for k in kn ])

        # Multiply transform of odd samples by twiddle factors:
        tXo = list(zip(W_real, W_imag, Xo_real, Xo_imag))
        WXo_real = array('f', [ Wkr*Xokr - Wki*Xoki for Wkr, Wki, Xokr, Xoki in tXo ])
        WXo_imag = array('f', [ Wkr*Xoki + Wki*Xokr for Wkr, Wki, Xokr, Xoki in tXo ])

        # Aggregate results:
        a_real = list(zip(Xe_real, WXo_real))
        a_imag = list(zip(Xe_imag, WXo_imag))
        X_real = ( array('f', [ Xekr + WXokr for Xekr, WXokr in a_real ]) +
                   array('f', [ Xekr - WXokr for Xekr, WXokr in a_real ]) )
        X_imag = ( array('f', [ Xeki + WXoki for Xeki, WXoki in a_imag ]) +
                   array('f', [ Xeki - WXoki for Xeki, WXoki in a_imag ]) )
        return [ Xr + 1j*Xi for Xr, Xi in zip(X_real, X_imag) ]


####################################################################################################
# Auxililary function to reorder a vector in bit-reversed order:
def bit_reverse(k, r):
    """
    Bit-reversed version of an integer number.

    :Parameters:
      k
        The number to be bit-reversed;
      r
        The number of bits to take into consideration when reversing.

    :Returns:
      The number k, bit-reversed according to integers with r bits.
    """
    l = 0                                      # Accumulate the results;
    for i in range(0, r):                      # Loop on every bit;
        l = (l << 1) + (k & 1)                 # Test less signficant bit and add;
        k = (k >> 1)                           # Test next bit;
    return l


####################################################################################################
# Iterative Decimation-in-time FFT, using lists:
def iterative_fft(x):
    """
    Fast Fourier Transform using an iterative in-place decimation in time algorithm. This has
    O(N log_2(N)) complexity, and since there are less function calls, it will probably be
    marginally faster than the recursive versions. This uses arrays from the standard Python
    library.

    :Parameters:
      x
        The vector of which the DFT will be computed. Given the nature of the implementation, there
        is no restriction on the size of the vector, although it will almost always be called with
        a power of two size to give a fair comparison.

    :Returns:
      A complex-number vector of the same size, with the coefficients of the DFT.
    """
    N = len(x)                                 # Length of vector;
    r = int(math.log2(N))                      # Number of bits;
    X_real = array('f', [ xr.real for xr in x ])
    X_imag = array('f', [ xi.imag for xi in x ])
    for k in range(0, N):
        l = bit_reverse(k, r)                  # Reorder the vector according to the
        X_real[l] = x[k].real                  #   bit-reversed order;
        X_imag[l] = x[k].imag

    step = 1                                   # Auxililary for computation of twiddle factors;
    for k in range(0, r):
        for l in range(0, N, 2*step):
            W_real = math.cos(-math.pi / step)         # Twiddle factors;
            W_imag = math.sin(-math.pi / step)
            Wkn_real = 1.
            Wkn_imag = 0.
            for n in range(0, step):
                p = l + n
                q = p + step

                # Recombine results:
                aux_real = Wkn_real*X_real[q] - Wkn_imag*X_imag[q]
                aux_imag = Wkn_real*X_imag[q] + Wkn_imag*X_real[q]
                X_real[q] = X_real[p] - aux_real
                X_imag[q] = X_imag[p] - aux_imag

                aux_real = 2*X_real[p] - X_real[q]
                aux_imag = 2*X_imag[p] - X_imag[q]
                X_real[p] = aux_real
                X_imag[p] = aux_imag

                # Update twiddle factors:
                aux_real = Wkn_real * W_real - Wkn_imag * W_imag
                aux_imag = Wkn_real * W_imag + Wkn_imag * W_real
                Wkn_real = aux_real
                Wkn_imag = aux_imag

        step = 2*step

    return [ Xr + 1j*Xi for Xr, Xi in zip(X_real, X_imag) ]


####################################################################################################
# Auxiliary function:
def __factor(n):
    """
    Smallest prime factor of a given number. If the argument is prime itself, then it is the return
    value.

    :Parameters:
      n
        Number to be inspected.

    :Returns:
      The smallest prime factor, or the number itself if it is already a prime.
    """
    rn = int(math.ceil(math.sqrt(n)))          # Search up to the square root of the number;
    for i in range(2, rn+1):
        if n%i == 0:                           # When remainder is zero, factor is found;
            return i
    return n


####################################################################################################
# Recursive FFT:
def recursive_nfft(x):
    """
    Fast Fourier Transform using a recursive decimation in time algorithm for vectors of length
    different from a power of 2. This has smaller complexity than the direct FT, though the exact
    value is difficult to compute. This implementation uses native Python arrays.

    :Parameters:
      x
        The vector of which the FFT will be computed. Its length must be a composite number, or else
        the computation will be defered to the direct FT, and there will be no efficiency gain.

    :Returns:
      A complex-number vector of the same size, with the coefficients of the DFT.
    """
    N = len(x)                                 # Length of the vector;
    N1 = __factor(N)                           # Find the smallest factor of the vector length;
    if N1 == N:                                # If the length is prime itself,
        return direct_ft(x)                    #    the transform is given by the direct form;
    else:
        N2 = N // N1                           # Decompose in two factors, N1 being prime;
        X_real = array('f', [ 0. ] * N)        # Accumulate the results;
        X_imag = array('f', [ 0. ] * N)
        W_real = math.cos(-2*math.pi/N)        # Twiddle factors;
        W_imag = math.sin(-2*math.pi/N)
        Wj_real = 1.
        Wj_imag = 0.
        for j in range(N1):                    # Compute every subsequence of size N2;

            Xj = recursive_nfft(x[j::N1])
            Xj_real = array('f', [ xr.real for xr in Xj ])
            Xj_imag = array('f', [ xi.imag for xi in Xj ])
            Wkj_real = 1.
            Wkj_imag = 0.
            for k in range(N):

                # Recombine results:
                aux_real = Xj_real[k%N2]*Wkj_real - Xj_imag[k%N2]*Wkj_imag
                aux_imag = Xj_real[k%N2]*Wkj_imag + Xj_imag[k%N2]*Wkj_real
                X_real[k] += aux_real
                X_imag[k] += aux_imag

                # Update twiddle factors:
                aux_real = Wkj_real*Wj_real - Wkj_imag*Wj_imag
                aux_imag = Wkj_real*Wj_imag + Wkj_imag*Wj_real
                Wkj_real = aux_real
                Wkj_imag = aux_imag

            aux_real = Wj_real*W_real - Wj_imag*W_imag
            aux_imag = Wj_real*W_imag + Wj_imag*W_real
            Wj_real = aux_real
            Wj_imag = aux_imag

        return [ Xr + 1j*Xi for Xr, Xi in zip(X_real, X_imag) ]


####################################################################################################
# Main program:
if __name__ == "__main__":
    pass
